### ISSUE TITLE
Add Type Hints Across the Entire BardGPT Codebase

### ISSUE THREAD

**Issue by Himanshu7921** [Himanshu7921/BardGPT#5]
## 1. Issue Type

* [ ] Bug Report
* [x] Feature Request
* [x] Documentation Improvement
* [ ] Training / Sampling Issue
* [ ] Performance / Optimization Request
* [ ] Other

---

## 2. Description

The BardGPT codebase currently lacks **consistent type hints** across most modules.
Adding type annotations will greatly improve:

* Code readability
* Contributor onboarding
* IDE auto-completion
* Static analysis
* Catching errors before runtime
* Overall maintainability

This issue proposes adding **full type hints** to every file.

Modules needing type annotations include:

* `model/` (attention, FFN, transformer blocks, decoder model)
* `utils/` (after future refactor)
* `data/` loader
* `sampling/` utils (top-k, top-p, generate)
* Training logic (`train.py`)
* Config and checkpoint utilities

---

## 3. Steps to Reproduce (for bugs)

*Not a bug — enhancement request.*

---

## 4. Environment Information

* OS: Any
* Python version: 3.10+ recommended
* PyTorch version: Any
* CUDA version: Optional
* GPU model: Optional
* BardGPT commit hash or version: Latest
* Installed dependencies: N/A

---

## 5. Additional Context

Examples of missing type hints:

### In model classes

```python
def forward(self, x, mask):
    ...
```

should be:

```python
def forward(self, x: torch.Tensor, mask: torch.Tensor) -> torch.Tensor:
```

### In DataLoader

```python
def get_batch(self, split='train', batch_size=64, device=torch.device):
```

should become:

```python
def get_batch(
    self,
    split: str,
    batch_size: int,
    device: torch.device
) -> tuple[torch.Tensor, torch.Tensor]:
```

### In utilities

Checkpoint functions, sampling functions, and masking functions should clearly annotate:

* Input tensor types
* Return shapes
* Optional arguments
* Callable signatures

This dramatically helps contributors working on architecture, debugging, or refactoring.

---

### **Contributor Tasks**

* [ ] Add type hints to all functions in `model/`
* [ ] Add type hints to DataLoader
* [ ] Add type hints to sampling utilities (top-k, top-p, generate)
* [ ] Add type hints to checkpoint utilities
* [ ] Add type hints to config classes
* [ ] Add type hints to training loops
* [ ] Add return type annotations everywhere
* [ ] Run `mypy` or `pyright` to validate correctness
* [ ] Update README with "Type Hints Used" note

### Optional Enhancements

* Add `pyproject.toml` with mypy configuration
* Add GitHub Actions workflow for type checks

---

Thank you for taking the time to report this issue.
Your contribution helps improve BardGPT for everyone.
